### ISSUE TITLE
Module.bind() Error When Calling Scala API from Java

### ISSUE THREAD

**Issue by nm25tunn** [apache/incubator-mxnet#7136]
## Environment info
Operating System: RHEL_64bit

Compiler: JDK 1.8

Package used (Python/R/Scala/Julia): Scala

MXNet version:  0.10

## Error Message:
```
java.lang.IllegalArgumentException: requirement failed: shape inference failed
	at scala.Predef$.require(Predef.scala:224) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at ml.dmlc.mxnet.module.DataParallelExecutorGroup.ml$dmlc$mxnet$module$DataParallelExecutorGroup$$bindIthExec(DataParallelExecutorGroup.scala:583) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at ml.dmlc.mxnet.module.DataParallelExecutorGroup$$anonfun$bindExec$1.apply(DataParallelExecutorGroup.scala:356) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at ml.dmlc.mxnet.module.DataParallelExecutorGroup$$anonfun$bindExec$1.apply(DataParallelExecutorGroup.scala:355) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at scala.collection.immutable.Range.foreach(Range.scala:160) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at scala.collection.AbstractTraversable.map(Traversable.scala:104) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at ml.dmlc.mxnet.module.DataParallelExecutorGroup.bindExec(DataParallelExecutorGroup.scala:355) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at ml.dmlc.mxnet.module.DataParallelExecutorGroup.<init>(DataParallelExecutorGroup.scala:317) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at ml.dmlc.mxnet.module.DataParallelExecutorGroup$Builder.build(DataParallelExecutorGroup.scala:223) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
	at ml.dmlc.mxnet.module.Module.bind(Module.scala:243) ~[mxnet-full_2.11-linux-x86_64-gpu-0.10.0-SNAPSHOT.jar:?]
```

## Minimum reproducible example
```java
    NDArray ndInput = NDArray.array(mapInput(history),
                new Shape(new ArrayBuffer<>().$plus$eq(1).$plus$eq(this.inMap.size())), Context.cpu(0));
    ArrayBuffer<NDArray> dataTemp = new ArrayBuffer<NDArray>().$plus$eq(ndInput);
    NDArrayIter data = new NDArrayIter(dataTemp, new ArrayBuffer<>(), 1, false, "discard", "data",  null);
    ---> this.model.bind(DataDesc.ListMap2Descs(data.provideData()),
            Option.apply(null), false, false, true, Option.apply(null), "write");
    this.model.initParams(new Uniform(0.01f), null, null, true, false);
    float[] out = this.model.predict(data, 1, true).head().toArray();
```

* The function call mapInput(history) returns a flat Java array of floating point values the same size as inMap.size() (I have verified this is working as I expect)
* The NDArray's dimensions should be 1 x inMap.size() (batch of 1 x the number of features).
*  this.model is the Module that was loaded with the Module.loadCheckpoint() method, I have verified the model loads from file properly
* ---> indicates the line in my code that the error is occurring on
* I am loading a model which was trained using the Python package
* The model relies on a custom loss function, I have re-implemented the loss function as a CustomOp that simply forwards the input to it along (no need to calculate loss here since my use case is only performing predictions, never back prop)

For context here is the loss:
```java
public class MyLoss extends CustomOp {

    @Override
    public void forward(final boolean isTrain, final String[] req, final NDArray[] inData, final NDArray[] outData,
                        final NDArray[] aux)
    {
        //No need to calculate loss, just copy predictions forward, which are in the first NDArray in inData
        assign(outData[0], req[0], inData[0]);
    }

    @Override
    public void backward(final String[] req, final NDArray[] outGrad, final NDArray[] inData, final NDArray[] outData,
                         final NDArray[] inGrad, final NDArray[] aux)
    {
        //Doesn't require implementation for forward prop
    }
}
```
And it's corresponding property:
```java
public class MyProp extends CustomOpProp {

    /**
     * Constructor for MyProp used to load model
     */
    public MyProp() {
        super(false);
    }

    @Override
    public CustomOp createOperator(final String ctx, final int[][] inShapes, final int[] inDtypes) {
        return new MyLoss();
    }

    @Override
    public Tuple3<Shape[], Shape[], Shape[]> inferShape(Shape[] inShape) {
        return new Tuple3<>(inShape, new Shape[]{inShape[0]}, new Shape[0]);
    }

    @Override
    public String[] listArguments() {
        return new String[]{"data", "label"};
    }

    @Override
    public String[] listOutputs() {
        return new String[]{"preds"};
    }

}
```

## What have you tried to solve it?

1. Giving the label a name "label" or "softmax_label" (has no effect, same stack trace)
2. Provide an empty array as a label (identical shape as the data), code crashes out earlier due to heap space running out (also doing this is non optimal due to the resource usage, unless absolutely necessary).
3. Tried switching the order of the NDArray's shape arguments (no effect)

I have a suspicion that it is tied to how I am setting up the iterator but I am not certain, if anyone has run into this issue or can identify a mistake in what I've written then I'd greatly appreciate it!
@javelinjs @Ldpe2G

**Comment by Ldpe2G** [comment]
@nm25tunn seems you did not provide enough input shapes information for the symbol to infer the network's shape.

**Comment by nm25tunn** [comment]
@Ldpe2G I'm not sure what you mean, is there supposed to be more dimensionality for my data array or does the label also need to be provided a shape (even if there is no label)?

**Comment by Ldpe2G** [comment]
@nm25tunn since I did not know the definition of your network, I can't tell what actually cause this problem.
Have you check the number of shapes you provide match the number of your network's input?

**Comment by Ldpe2G** [comment]
Or you could use the symbol's bind method to locate the problem.

**Comment by nm25tunn** [comment]
I modified the code to pass a label (populated with 0s and equivalent shape to the output layer) and now it makes it past shape inference but it segfaults on type inference.  I passed the jvmarg -Xcheck:jni the next time I ran to check the jni calls and I am getting the following stack trace.

```
     [java] FATAL ERROR in native method: Non-array passed to JNI array operations
     [java]     at ml.dmlc.mxnet.LibInfo.mxSymbolInferType(Native Method)
     [java]     at ml.dmlc.mxnet.Symbol.inferType(Symbol.scala:194)
     [java]     at ml.dmlc.mxnet.Symbol.inferType(Symbol.scala:185)
     [java]     at ml.dmlc.mxnet.module.DataParallelExecutorGroup.ml$dmlc$mxnet$module$DataParallelExecutorGroup$$bindIthExec(DataParallelExecutorGroup.scala:588)
     [java]     at ml.dmlc.mxnet.module.DataParallelExecutorGroup$$anonfun$bindExec$1.apply(DataParallelExecutorGroup.scala:356)
     [java]     at ml.dmlc.mxnet.module.DataParallelExecutorGroup$$anonfun$bindExec$1.apply(DataParallelExecutorGroup.scala:355)
     [java]     at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
     [java]     at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
     [java]     at scala.collection.immutable.Range.foreach(Range.scala:160)
     [java]     at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
     [java]     at scala.collection.AbstractTraversable.map(Traversable.scala:104)
     [java]     at ml.dmlc.mxnet.module.DataParallelExecutorGroup.bindExec(DataParallelExecutorGroup.scala:355)
     [java]     at ml.dmlc.mxnet.module.DataParallelExecutorGroup.<init>(DataParallelExecutorGroup.scala:317)
     [java]     at ml.dmlc.mxnet.module.DataParallelExecutorGroup$Builder.build(DataParallelExecutorGroup.scala:223)
     [java]     at ml.dmlc.mxnet.module.Module.bind(Module.scala:243)
```

**Comment by nm25tunn** [comment]
Resolved this by defining the following in my CustomOpProp:

```java
@Override
    public Tuple3<Enumeration.Value[], Enumeration.Value[], Enumeration.Value[]> inferType(Enumeration.Value[] inType) {
        return new Tuple3<>(inType, new Enumeration.Value[]{inType[0]}, new Enumeration.Value[0]);
    }
```
Although now I am getting a failure on the predict() statement:
```
c_mxnet_native_c_api.cc:2165: Check failed: e == cudaSuccess || e == cudaErrorCudartUnloading CUDA: no CUDA-capable device is detected
```
Although I'm not sure why I'm getting this since I have defined the context as Context.cpu(0) throughout.  I will update this post with a solution if I find one.

**Comment by Ldpe2G** [comment]
@nm25tunn which mxnet version are you use?

**Comment by nm25tunn** [comment]
@Ldpe2G I'm using MxNet 0.10.0

**Comment by Ldpe2G** [comment]
@nm25tunn it's a bit strange, the error is caused by this line:
https://github.com/Ldpe2G/mxnet/blob/bbad27c5a5654298e9c6616442c46054b36a7307/scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc#L2165

If you compile mxnet without cuda, it should not be enter into this code block.

**Comment by nm25tunn** [comment]
@Ldpe2G I haven't had a chance to recompile without CUDA yet but in the meantime I moved the code to a machine with a GPU and it works as expected there now.  Is there a reason from an architectural standpoint that the package is attempting to initialize a GPU or is this a bug?

**Comment by Ldpe2G** [comment]
@nm25tunn You can think of it as a bug. May be I should check whether the number of gpu is greater than zero before I set the cuda device.  Or maybe you can make a pr to fix this :) .

**Comment by nm25tunn** [comment]
@Ldpe2G I can do that for sure when I have a bit more bandwidth, I should be able to have a fix in the next few weeks.

**Comment by Ldpe2G** [comment]
@nm25tunn great, looking forward to your pr :)

**Comment by mg0880gm** [comment]
@Ldpe2G Similar issue as @nm25tunn mentioned when run MxNet on a host without GPU. I submitted a pull request about proposal to fix this and the link: https://github.com/apache/incubator-mxnet/pull/7855. Could you help review in some convenient time?

Basically the proposal is to call cudaSetDevice only when the input Context is not CPU. In local evaluation this help fix the issue successfully in a CPU-only host.

**Comment by nikrao** [comment]
@nm25tunn @mg0880gm was this issue resolved and the fix merged?

**Comment by mg0880gm** [comment]
The related pr has been merged: https://github.com/apache/incubator-mxnet/pull/7855. The issue has been proved to be fixed. Please kindly help resolve this.
